packages/flutter/lib/src/widgets/localizations.dart

```diff
@@ -0,0 +1,459 @@
+/// asynchronously. For example localized resources might be loaded from
+/// the nextwork. The [load] method returns a Future that completes
+/// when all of the loaders have completed.
+///
+/// See also:
+///
+///  * [WidgetsApp.localizationsDelegate] and [MaterialApp.localizationsDelegate],
+///    which enable configuring the app's LocalizationsLoader.
+class DefaultLocalizationsDelegate extends LocalizationsDelegate {
+  DefaultLocalizationsDelegate(this.allLoaders) {
+    assert(allLoaders != null);
+  }
+
+  /// The [LocalizationsLoader]s in this map define all of the collections
+  /// of resources for a [Localizations] widget.
+  ///
+  /// The [load] method's Future completes when all of the loaders' load functions
+  /// have completed.
+  final Map<Type, LocalizationsLoader> allLoaders;
+
+  final Map<Locale, Map<Type, dynamic>> _localeToResources = <Locale, Map<Type, dynamic>>{};
+  final Set<Locale> _loading = new Set<Locale>();
+
+  @override
+  Future<Iterable<Type>> load(Locale locale) {
+    assert(locale != null);
+    assert(!_loading.contains(locale));
+
+    Iterable<Type> types;
+
+    _loading.add(locale);
+    final Future<Iterable<Type>> typesFuture = _loadMap<Type, dynamic>(
+      new Map<Type, Future<dynamic>>.fromIterables(
+        allLoaders.keys,
+        allLoaders.values.map<Future<dynamic>>((LocalizationsLoader loader) => loader(locale)),
+      ),
+    ).then<Iterable<Type>>((Map<Type, dynamic> resources) {
+      _loading.remove(locale);
+      _localeToResources[locale] = resources;
+      return types = resources.keys;
+    });
+
+    return types == null ? typesFuture : new SynchronousFuture<Iterable<Type>>(types);
+  }
+
+  @override
+  LocalizationsStatus statusFor(Locale locale) {
+    if (_localeToResources[locale] != null)
+      return LocalizationsStatus.loaded;
+    if (_loading.contains(locale))
+      return LocalizationsStatus.loading;
+    return LocalizationsStatus.none;
+  }
+
+  @override
+  T resourcesFor<T>(Locale locale, Type type) {
+    assert(locale != null);
+    assert(type != null);
+    final Map<Type, dynamic> resources = _localeToResources[locale];
+    if (resources == null)
+      return null;
+    // TODO(hansmuller) assert(resources[type] is type) when
+    // https://github.com/dart-lang/sdk/issues/27680 has been resolved.
+    return resources[type];
+  }
+
+  @override
+  bool updateShouldNotify(DefaultLocalizationsDelegate old) => false;
+}
+
+/// Create single [LocalizationsDelegate] from a list of them.
+///
+/// All of the delegates' load methods must return disjoint lists of types, i.e.
+/// only one delegate's [resourcesFor] method can be responsible for a set of
+/// localizations of a particular [Type].
+///
+/// When possible, using DefaultLocalizationsDelegate should be preferred to
+/// using this class since it's one less level of indirection.
+class MergedLocalizationsDelegate extends LocalizationsDelegate {
```

Comment:
To simplify the API (though it doesn't do much to the implementation), I'd have a `LocalizationsDelegate.merge` factory constructor that returns an instance of this class (which can then be made private).